File: go/data-structures/easy/dynamicArray.go
/*
Author: Rithie Natan Carvalhaes Prado
Date: 2026-06-14
Last modified: 2026-06-14
Description: Dynamic Array - HackerRank
See: https://www.hackerrank.com/challenges/dynamic-array/problem
*/
//--------- package ---------
package easy


//--------- functions ---------
// dynamicArray - Processes a list of queries on a dynamic array and returns the results.
func dynamicArray(n int32, queries [][]int32) []int32 {
	var lastAnswer int32 = 0;
	var ans []int32 = []int32{};

	var arr [][]int32 = make([][]int32, n);

	for i := 0; i < len(queries); i++ {
		if queries[i][0] == 1 {
			arr[(queries[i][1] ^ lastAnswer) % n] = append(arr[(queries[i][1] ^ lastAnswer) % n], queries[i][2]);
		} else {
			lastAnswer = arr[(queries[i][1] ^ lastAnswer) % n][queries[i][2]%int32(len(arr[(queries[i][1] ^ lastAnswer) % n]))]
			ans = append(ans, lastAnswer);
		}
	}

	return ans;
}//end dynamicArray()

File: python/data-structures/easy/dynamicArray.py
"""
Author: Rithie Natan Carvalhaes Prado
Date: 2026-06-14
Last modified: 2026-06-14
Description: Dynamic Array - HackerRank
Test description: https://www.hackerrank.com/challenges/dynamic-array/problem
"""
#---------- functions ----------
def dynamicArray(n: int, queries: list[list[int]]) -> list[int]:
    """
    Process a list of queries on a dynamic array and return the results.

    Parameters:
        n (int): The number of sequences in the dynamic array.
        queries (list[list[int]]): A list of queries to process.

    Returns:
        list[int]: A list of the results of the queries.
    """
    lastAnswer: int = 0
    ans: list[int] = []
    arr: list[list[int]] = [[] for _ in range(n)]

    for q in queries:
        if q[0] == 1:
            idx: int = (q[1] ^ lastAnswer) % n
            arr[idx].append(q[2])
        else:
            idx: int = (q[1] ^ lastAnswer) % n
            lastAnswer = arr[idx][q[2] % len(arr[idx])]
            ans.append(lastAnswer)

    return ans